src/lib/consensus/global_slot.ml

```diff
@@ -73,18 +74,28 @@ module Make_str (A : Wire_types.Concrete) = struct
 
   let slots_per_epoch { Poly.slots_per_epoch; _ } = slots_per_epoch
 
-  let epoch (t : t) = UInt32.Infix.(t.slot_number / t.slots_per_epoch)
+  let epoch (t : t) =
+    let slot_number_u32 = T.to_uint32 t.slot_number in
+    UInt32.Infix.(slot_number_u32 / t.slots_per_epoch)
 
-  let slot (t : t) = UInt32.Infix.(t.slot_number mod t.slots_per_epoch)
+  let slot (t : t) =
+    let slot_number_u32 = T.to_uint32 t.slot_number in
+    UInt32.Infix.(slot_number_u32 mod t.slots_per_epoch)
 
   let to_epoch_and_slot t = (epoch t, slot t)
 
   let ( + ) (x : t) n : t =
-    { x with slot_number = T.add x.slot_number (T.of_int n) }
+    { x with
+      slot_number = T.add x.slot_number (Mina_numbers.Global_slot_span.of_int n)
+    }
+
+  let ( < ) (t : t) (t' : t) = T.compare t.slot_number t'.slot_number < 0
 
-  let ( < ) (t : t) (t' : t) = UInt32.compare t.slot_number t'.slot_number < 0
+  let ( - ) (t : t) (t' : t) =
+    T.sub t.slot_number
+      (Mina_numbers.Global_slot_span.of_uint32 @@ T.to_uint32 t'.slot_number)
```

Comment:
I'm not sure I get why we need the conversion for `t'` and not for `t` as they have the same type so I was expecting this treatment to be symmetric (old code had no conversion whatsoever)

Reply:
`T.sub` takes a slot and a span, so `t'.slot_number` needed to be converted to a span.

`T.sub` used to take a slot and another slot, but I think it makes more sense to subtract a span, in the same way `Time` does.

But `( - )` isn't used anywhere, so I'm going to remove it.  Its uses were replaced by `diff_slots`, which takes two slots and returns a span.